### PR TITLE
Introducing Email template internationalization Feature

### PR DESCRIPTION
**Pull request by kasungayan**

### CODE REVIEW

**Comment by kasungayan** [comment]
Adding ability to add multiple locales to email templates
Adding the ability to view multi locale templates
Modifying existing registry persistence methodology for email templates (Moving email template content to "Resource" artifact from "Property" artifact)
Adding Listener on Tenant creation when loading email templates
Rectifying existing limitations of Email saving Functionality

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt.ui/pom.xml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            <artifactId>org.wso2.carbon.logging</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.identity</groupId>
+            <artifactId>org.wso2.carbon.i18n.mgt.stub</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.orbit.org.owasp.encoder</groupId>
+            <artifactId>encoder</artifactId>
+            <version>1.2.0.wso2v1</version>
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt.ui/pom.xml]: Move the version to the repo's parent pom

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt.ui/src/main/resources/web/i18n-mgt/add-email-template-submit.jsp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt.ui/src/main/resources/web/i18n-mgt/add-email-template-submit.jsp]: remove these unnecessary lines

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/Config.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+public interface Config {
+	
+	String getProperty(String key);
+
+	void setProperty(String key, String value);
+
+	Properties getProperties();
+
+	void setProperties(Properties props);
+}
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/Config.java]: replace tab character with 4 spaces in this file

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigBuilder.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+				String configType2 = omElement.getAttributeValue(new QName(
+						"prop"));
+				String configFinal = configType1 + "|" + configType2;
+				if (StringUtils.isNotBlank(configFinal)) {
+					emailConfig.setProperty(configFinal, loadEmailConfig(omElement));
+				}
+
+			}
+		} catch (XMLStreamException e) {
+			log.warn("Error while loading email config. using default configuration");
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigBuilder.java]: log 'e'

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigBuilder.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+				emailTemplate.append("|");
+				emailTemplate.append(element.getText());
+			} else if ("footer".equals(element.getLocalName())) {
+				emailTemplate.append("|");
+				emailTemplate.append(element.getText());
+			}
+		}
+		return emailTemplate.toString();
+	}
+}
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigBuilder.java]: replace tab character with 4 spaces in this file

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigBuilder.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+				String configFinal = configType1 + "|" + configType2;
+				if (StringUtils.isNotBlank(configFinal)) {
+					emailConfig.setProperty(configFinal, loadEmailConfig(omElement));
+				}
+
+			}
+		} catch (XMLStreamException e) {
+			log.warn("Error while loading email config. using default configuration");
+		} catch (FileNotFoundException e) {
+			log.warn("Error while loading email config. using default configuration");
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigBuilder.java]: log 'e'

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigBuilder.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+		} finally {
+			try {
+				if (parser != null) {
+					parser.close();
+				}
+				if (stream != null) {
+					stream.close();
+				}
+			} catch (XMLStreamException e) {
+				log.error("Error while closing XML stream");
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigBuilder.java]: log 'e'

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigBuilder.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+			} catch (IOException e) {
+				log.error("Error while closing input stream");
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigBuilder.java]: log 'e'

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigReader.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+	/**
+	 * Reads a tenant specific configuration
+	 *
+	 * @param tenantId     -  Tenant Id which the Email Template configuration belongs.
+	 * @param resourcePath - Resource path to be saved the email templates.
+	 * @return The populated email template configuration properties.
+	 * @throws org.wso2.carbon.i18n.mgt.exceptions.I18nMgtEmailConfigException
+	 */
+	Properties read(int tenantId, String resourcePath) throws I18nMgtEmailConfigException;
+}
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigReader.java]: replace tab character with 4 spaces in this file

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigType.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.i18n.mgt.config;
+
+public enum ConfigType {
+	EMAIL
+}
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigType.java]: replace tab character with 4 spaces in this file

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigWriter.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+	/**
+	 * Writes configuration to a tenant specific persistence store.
+	 *
+	 * @param tenantId     - Tenant Id which the email template configuration belongs.
+	 * @param props        - Email properties to be saved.
+	 * @param resourcePath - Resource path to be saved the email templates.
+	 * @throws org.wso2.carbon.i18n.mgt.exceptions.I18nMgtEmailConfigException
+	 */
+	void write(int tenantId, Properties props, String resourcePath) throws I18nMgtEmailConfigException;
+}
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigWriter.java]: replace tab character with 4 spaces in this file

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/EmailConfigTransformer.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+	public static Properties transform(EmailTemplateDTO template) throws IdentityException {
+		Properties props = new Properties();
+		StringBuilder contents = new StringBuilder();
+		contents.append(template.getSubject()).append("|").append(template.getBody()).append("|").
+				append(template.getFooter());
+		props.setProperty(template.getName(), contents.toString());
+		return props;
+	}
+}
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/EmailConfigTransformer.java]: replace tab character with 4 spaces in this file

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/EmailNotificationConfig.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+		return this.properties;
+	}
+
+	@Override
+	public void setProperties(Properties props) {
+		this.properties = props;		
+	}
+	
+	
+}
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/EmailNotificationConfig.java]: replace tab character with 4 spaces in this file

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/RegistryConfigAdd.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+                            + " " + "to resource location : " + " " + emailTemplateFileResourcePath);
+                }
+
+            } else {
+                throw new I18nMgtEmailConfigException("Resource duplication error while adding email template : "
+                        + " " + emailTemplateFilename);
+            }
+        } catch (RegistryException e) {
+            throw new I18nMgtEmailConfigException("Error occurred while adding email template to registry path : "
+                    + resourcePath);
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/RegistryConfigAdd.java]: 'e' is not thrown

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/RegistryConfigReader.java [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+			if (log.isDebugEnabled()) {
+				log.debug("Ignored ResourceNotFoundException", e);
+			}
+		} catch (RegistryException e) {
+			throw new I18nMgtEmailConfigException(
+					"Error occurred while reading email templates from path : " + resourcePath, e);
+		}
+		return readerProps;
+	}
+}
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/RegistryConfigReader.java]: replace tab character with 4 spaces in this file

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/RegistryConfigWriter.java [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+				String[] propsKeyName = names.toArray(new String[names.size()]);
+				String[] propsKeyNameSplit = propsKeyName[0].split("\\.");
+				String emailTemplateFolderName = propsKeyNameSplit[0];
+				String emailTemplateFilename = propsKeyName[0];
+				String emailTemplateFileResourcePath = resourcePath + I18nMgtConstants.EMAIL_FOLDER_SEPARATOR +
+						emailTemplateFolderName + I18nMgtConstants.EMAIL_FOLDER_SEPARATOR + emailTemplateFilename;
+
+				Resource resource = userReg.newResource();
+				String emailTemplateContent = props.getProperty(emailTemplateFilename);
+				resource.setMediaType("text/plain");
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/RegistryConfigWriter.java]: Use constants

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/RegistryConfigWriter.java [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+				userReg.put(emailTemplateFileResourcePath, resource);
+
+				if (log.isDebugEnabled()) {
+					log.debug("Email template : " + emailTemplateFilename + " " +
+							"saved successfully to path " + " " + emailTemplateFileResourcePath);
+				}
+
+			}
+		} catch (RegistryException e) {
+			throw new I18nMgtEmailConfigException("Error occurred while adding email template to registry path : "
```
Comment (pulasthi7) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/RegistryConfigWriter.java]: e is not thrown

Comment (kasungayan) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/RegistryConfigWriter.java]: Fixed with https://github.com/kasungayan/carbon-identity/commit/026d6476d4e81b9b7a14ae5e2fbb45136e498a9f

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt.ui/pom.xml [hunk 18]
```diff
@@ -0,0 +1,79 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (omindu) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt.ui/pom.xml]: Missing licence header

Comment (kasungayan) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt.ui/pom.xml]: Fixed with ab38236f41939624e097f157bd4df46d77167902

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt.ui/src/main/java/org/wso2/carbon/i18n/mgt/ui/I18nEmailMgtConfigServiceClient.java [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+    private static Log log = LogFactory.getLog(I18nEmailMgtConfigServiceClient.class);
+
+    private I18NEmailMgtConfigServiceStub stub;
+
+    public I18nEmailMgtConfigServiceClient(String url,
+                                           ConfigurationContext configContext) throws Exception {
+        try {
+            stub = new I18NEmailMgtConfigServiceStub(configContext, url
+                    + "I18nEmailMgtConfigService");
+        } catch (java.lang.Exception e) {
```
Comment (prabathabey) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt.ui/src/main/java/org/wso2/carbon/i18n/mgt/ui/I18nEmailMgtConfigServiceClient.java]: package name of "Exception" is redundant. You can simply use "Exception" instead.

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt.ui/src/main/java/org/wso2/carbon/i18n/mgt/ui/I18nEmailMgtConfigServiceClient.java [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+        try {
+            stub = new I18NEmailMgtConfigServiceStub(configContext, url
+                    + "I18nEmailMgtConfigService");
+            ServiceClient client = stub._getServiceClient();
+            Options option = client.getOptions();
+            option.setManageSession(true);
+            option.setProperty(
+                    org.apache.axis2.transport.http.HTTPConstants.COOKIE_STRING,
+                    cookie);
+        } catch (java.lang.Exception e) {
```
Comment (prabathabey) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt.ui/src/main/java/org/wso2/carbon/i18n/mgt/ui/I18nEmailMgtConfigServiceClient.java]: Same goes here as well.

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/Config.java [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.i18n.mgt.config;
+
+import java.util.Properties;
+
+public interface Config {
```
Comment (prabathabey) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/Config.java]: No comments added to either Class or method definitions. Please add some meaningful comments explaining the semantics of this particular interface and its methods.

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/EmailNotificationConfig.java [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+package org.wso2.carbon.i18n.mgt.config;
+
+
+import java.util.Properties;
+
+public class EmailNotificationConfig implements Config {
+
+    private Properties properties;
+
+
```
Comment (prabathabey) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/EmailNotificationConfig.java]: Unnecessary new line.

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/EmailNotificationConfig.java [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+    public EmailNotificationConfig() {
+        properties = new Properties();
+    }
+
+    @Override
+    public String getProperty(String key) {
+
```
Comment (prabathabey) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/EmailNotificationConfig.java]: Unnecessary new line.

File: components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigBuilder.java [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+						if (config == null) {
+							throw new I18nMgtEmailConfigException("Default email templates haven't written successfully");
+						}
+						break;
+					default:
+						throw new I18nMgtEmailConfigException("Configuration type not supported");
+				}
+				break;
+			case DB:
+				break;
```
Comment (prabathabey) [inline components/identity-i18n-mgt/org.wso2.carbon.i18n.mgt/src/main/java/org/wso2/carbon/i18n/mgt/config/ConfigBuilder.java]: Redundant "break" statement.

**Pull request closed without merging**